source/templates.tex

```diff
@@ -3160,7 +3160,14 @@
 be used in such a declaration.
 
 \pnum
-A non-template friend declaration shall not have a \grammarterm{requires-clause}.
+A non-template friend declaration
+with a \grammarterm{requires-clause}
+shall be a definition.
+A friend function template
+with a constraint that depends on a template parameter from an enclosing template
+shall be a definition.
+Such a constrained friend function or function template declaration
+does not declare the same function or function template as a declaration in any other scope.
```

Comment:
How is "in any _other_ scope" to be interpreted here? [[class.friend]/7](https://eel.is/c++draft/class.friend#7) mentions lexical scope (of the class) for hidden friend definitions, but namespace scope, as per [[class.friend]/6](https://eel.is/c++draft/class.friend#6), for the function itself.

As per [[over.dcl]/1](https://eel.is/c++draft/over.dcl#1) two function declarations refer to the same function only if they have the same trailing requires-clauses (if any). This should arguably also apply for non-template hidden friend definitions which differs only in their trailing requires-clauses. The non-normative text of [over.dcl]/1 contains an example for this particular case but for non-template member functions as compared to hidden friends.

This seems to have a few compilers confused, at least, as GCC, Clang and MSVC all (using current HEAD) implement different behaviours for overloading hidden friends only by different trailing requires-clauses:

```
struct Base {};

template<int N>
struct S : public Base {
    friend int f(Base&) requires (N == 1) { return 1; }
    friend int f(Base&) requires (N == 2) { return 3; }
    // Separate functions (which overload resolution will
    // have a hard time to disambiguate)?
};

int main() {
    S<1> s1{};
    S<2> s2{};    // GCC error: redefinition 
    (void)f(s1);  // MSVC error: ambiguous call (correct?)
    (void)f(s2);
    // Clang error: definition with same mangled name
    //              '_Z3fooR4Base' as another definition
}
```

Is MSVC correct here?

Should I be filing compiler bug reports for this, or should the normative text possibly be updated to be more explicit regarding this special case?

Reply:
I think MSVC is correct here. With respect to the Clang behaviour, the description of the signature indicates that the mangling should be unique. With respect to the GCC behaviour, the "does not declare" wording is meant to say that the friends declared by each specialization is unique.

Reply:
Hubert, thanks for the answer and the clarification; it's clear that Clang is wrong by violation of [[defns.signature.friend]](https://eel.is/c++draft/defns.signature.friend), and that GCC is wrong as per the intended meaning of [temp.friend]/9.

However, I'm wondering what is the actual correct behaviour is here; is MSVC really correct, or is "correct" even well-specified?

I'm having a hard time finding the normative reference for constraints checking when finding _viable_ candidates in overload resolution for the case of overloaded, constrained hidden friends, where the signatures solely differ by means of the their (mutually exclusive) trailing requires-clauses. Particularly, whether and how constraints checking will be performed during overload resolution on a free (friend) non-template function.

Starting with lookup, [[basic.lookup.argdep]/4.3](https://eel.is/c++draft/basic.lookup.argdep#4.3) states that any (namespace-scope) friend function
> [...] declared in classes with reachable definitions in the set of associated entities are visible [...]

can be found via ADL on associated entities. Thus, as per [[basic.lookup.argdep]/2.2](https://eel.is/c++draft/basic.lookup.argdep#2.2), for the calls `f(s1)` and the `f(s2)` in the example above the associated entity is the class template itself, even if the arguments are objects of particular specializations of the class template. This would arguably mean that ADL, for both calls `f(s1)` and `f(s2)`, can find both overloads as candidate functions? So far, this is the interpretation MSVC has done, as it proceeds (and subsequently fails) with overload resolution.

For finding viable functions during overload resolution, [[over.match.viable]/3](https://eel.is/c++draft/over.match.viable#3) applies:
> Second, for a function to be viable, if it has associated constraints ([temp.constr.decl]), those constraints shall be satisfied ([temp.constr.constr]).

but the case of hidden friends with associated constraints is special in this regard. In say the call `f(s1)`, how would constraint satisfaction be checked? At this point, given that ADL has actually found both candidates (as is the interpretation of MSVC), the constraint is in no apparent way connected to the particular argument (which is an object of the specialization which uniquely defined one of the overloads).

Intuitively, the only plausible variations are that

 - (A) constraints satisfaction cannot evaluated, and both candidate functions should be considered non-viable, or that
 - (B) ADL should never find a constrained hidden friend as a candidate function for lookup on a specialization for which the friend's constraint is not satisfied.

(A) would result in the program above being ill-formed (no viable function as constraints cannot be checked for satisfaction), but this would generalize also to non-overloaded hidden friends and thus possibly not be applicable.

This leads us to (B), which would make the program above well-formed by only finding candidate functions whose constraints are actually satisfied, but afaict constraint satisfaction is not part of finding candidates, but only of the later step of rejecting candidates when create the set of viable candidates.

None of (A) or (B) are implemented by MSVC, which seemingly accepts both overloads as viable and fails due to ambiguity when attempting to find the best viable function (no constraints checking?).

Reply:
> Starting with lookup, [[basic.lookup.argdep]/4.3](https://eel.is/c++draft/basic.lookup.argdep#4.3) states that any (namespace-scope) friend function
> 
> > [...] declared in classes with reachable definitions in the set of associated entities are visible [...]
> 
> can be found via ADL on associated entities. Thus, as per [[basic.lookup.argdep]/2.2](https://eel.is/c++draft/basic.lookup.argdep#2.2), for the calls `f(s1)` and the `f(s2)` in the example above the associated entity is the class template itself, even if the arguments are objects of particular specializations of the class template. This would arguably mean that ADL, for both calls `f(s1)` and `f(s2)`, can find both overloads as candidate functions? So far, this is the interpretation MSVC has done, as it proceeds (and subsequently fails) with overload resolution.

Yes, it seems it should have succeeded because only one of the candidate functions is viable in each lookup result.

> In say the call `f(s1)`, how would constraint satisfaction be checked?

The constraint satisfaction can be checked in the context of the template instantiation.